### PR TITLE
Land Bigtable v2

### PR DESCRIPTION
**Pull request by tseaver**
This is the monster.  See https://github.com/GoogleCloudPlatform/gcloud-python/issues/1850#issuecomment-228192275 for the PRs which landed on the `bigtable-v2` branch.

### CODE REVIEW

**Comment by daspecster** [comment]
Looks like Appveyor is having the same issue that was reported earlier.
https://github.com/GoogleCloudPlatform/gcloud-python/issues/1933
https://github.com/grpc/grpc/issues/7135

@tseaver this PR is brining my browser to it's knees. Could you squash this down? (not sure that will help or not)

**Comment by lesv** [comment]
Tres we are ready for this to be published assuming you have all the required approvals.

**Comment by lesv** [comment]
@tseaver We are ready for this to be published assuming you have all the required approvals.

**Comment by tseaver** [comment]
@daspecster Honestly, squashing would *hurt*.  Might be better off scanning the merged-to-`bigtable-v2` PRs listed on the [tracking issue checklist](https://github.com/GoogleCloudPlatform/gcloud-python/issues/1850#issuecomment-228192275).

**Comment by daspecster** [comment]
@tseaver, ok sounds like a plan! Thanks!

**Comment by tseaver** [comment]
Just FTR, I was able to run the `bigtable` system tests under Python 3.4:

```bash
$ tox -re system-tests3 --notest
$ .tox/system-tests3/bin/pip install grpcio 
$ .tox/system-tests3/bin/python system_tests/run_system_test.py --package=bigtable
D0629 13:53:51.485668817   28141 ev_posix.c:101]             Using polling engine: poll
test_read_row (bigtable.TestDataAPI) ... ok
test_read_rows (bigtable.TestDataAPI) ... ok
test_read_with_label_applied (bigtable.TestDataAPI) ... ok
test_create_instance (bigtable.TestInstanceAdminAPI) ... ok
test_list_instances (bigtable.TestInstanceAdminAPI) ... ok
test_reload (bigtable.TestInstanceAdminAPI) ... ok
test_update (bigtable.TestInstanceAdminAPI) ... ok
test_create_column_family (bigtable.TestTableAdminAPI) ... ok
test_create_table (bigtable.TestTableAdminAPI) ... ok
test_delete_column_family (bigtable.TestTableAdminAPI) ... ok
test_list_tables (bigtable.TestTableAdminAPI) ... ok
test_update_column_family (bigtable.TestTableAdminAPI) ... ok
(grpcio shutdown noise elided)

----------------------------------------------------------------------
Ran 12 tests in 29.129s

OK
```

The `bigtable-happybase` system tests aren't yet Py3k-clean (they use native string literals for row keys, etc.)

**Comment by mbrukman** [comment]
What's the status of this PR? @daspecster, are you reviewing it?

**Comment by daspecster** [comment]
@mbrukman, I am looking it over.
I'm sure we need more eyes than mine on it :)

File: docs/bigtable-instance-api.rst [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+List Intances
+-------------
+
+If you want a comprehensive list of all existing intances, make a
+`ListInstances`_ API request with
+:meth:`Client.list_intances() <gcloud.bigtable.client.Client.list_intances>`:
+
+.. code:: python
+
+    intances = client.list_intances()
```
Comment (daspecster) [inline docs/bigtable-instance-api.rst]: s/intances/instances/g

File: docs/bigtable-instance-api.rst [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+.. code:: python
+
+    instance = client.instance(instance_id, display_name=display_name)
+
+``display_name`` is optional. When not provided,
+``display_name`` defaults to the ``instance_id`` value.
+
+Even if this :class:`Instance <gcloud.bigtable.instance.Instance>` already
+has been created with the API, you'll want this object to use as a
```
Comment (daspecster) [inline docs/bigtable-instance-api.rst]: "this object to use as a parent"  maybe should be "this object to use a parent"

**Comment by lesv** [comment]
@tswast PTAL

**Comment by lesv** [comment]
@jonparrott PTAL - could you review this PR - ideally, we'd like to get it published today as part of Bigtable GA.

File: docs/bigtable-instance-api.rst (hunk 2 above)
Comment (tseaver) [inline docs/bigtable-instance-api.rst]: f34f5c3

File: docs/bigtable-instance-api.rst (hunk 1 above)
Comment (tseaver) [inline docs/bigtable-instance-api.rst]: d920487

File: gcloud/bigtable/client.py [hunk 3]
```diff
@@ -18,8 +18,8 @@
 
 In the hierarchy of API concepts
 
-* a :class:`Client` owns a :class:`.Cluster`
-* a :class:`.Cluster` owns a :class:`Table <gcloud.bigtable.table.Table>`
+* a :class:`Client` owns a :class:`.Instance`
```
Comment (tswast) [inline gcloud/bigtable/client.py]: [nit] owns an Instance

File: gcloud/bigtable/client.py [hunk 4]
```diff
@@ -18,8 +18,8 @@
...
+* a :class:`.Instance` owns a :class:`Table <gcloud.bigtable.table.Table>`
```
Comment (tswast) [inline gcloud/bigtable/client.py]: [nit] an Instance

File: gcloud/bigtable/instance.py [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+    operation_id = int(match.group('operation_id'))
+
+    request_metadata = _parse_pb_any_to_native(operation_pb.metadata)
+    operation_begin = _pb_timestamp_to_datetime(
+        request_metadata.request_time)
+
+    return operation_id, location_id, operation_begin
+
+
+class Operation(object):
```
Comment (tswast) [inline gcloud/bigtable/instance.py]: I'm curious. Was this auto-generated?

Seems painful to have to write something like this for every object type.

File: gcloud/bigtable/client.py (hunk 4 above)
Comment (tseaver) [inline gcloud/bigtable/client.py]: Thanks!  055d2a2

File: gcloud/bigtable/client.py (hunk 3 above)
Comment (tseaver) [inline gcloud/bigtable/client.py]: Thanks!  055d2a2

**Comment by tswast** [comment]
Changes LGTM. I tested with this branch on my Hello World samples, and both ran fine. Thanks!

File: gcloud/bigtable/instance.py (hunk 5 above)
Comment (tseaver) [inline gcloud/bigtable/instance.py]: @tswast 

> I'm curious. Was this auto-generated?

Nope.  IT was copied and lightly modified from the one used for long-running V1 cluster operations.

> Seems painful to have to write something like this for every object type.

A bit.  Most of the issue is mapping between the values returned from the initial request and the values required to poll the operation status.  E.g., for a `CreateInstance` request, the response contains an `Operation` proto whose name/path includes the project and instance ID:

```
operations/projects/<PROJECT>/instances/<INSTANCE-ID>/locations/<LOCATION-ID>/operations/<OPERATION-ID>
```

but the value needed in the status-checking request isn't supposed to have them:

```
operations/locations/<LOCATION-ID>/operations/<OPERATION-ID>
```

The version for V2 `Cluster` is similar.

**Comment by tseaver** [comment]
@dhermes any objections to a merge?

**Comment by lesv** [comment]
@tseaver - We've had two pythonistas review. dhermes is away.  It's ok to merge.

**Comment by lesv** [comment]
@tseaver Can you let us know the PyPi status -- (ie submit and let us know)

**Comment by tseaver** [comment]
@lesv 0.17.0 is tagged.  Travis will push the release when the [tag build is finished](https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/141209279).

**Comment by lesv** [comment]
@tseaver  What about docs?

**Comment by tseaver** [comment]
@lesv [RTD](http://gcloud.readthedocs.io/en/stable/bigtable-instance-api.html) already shows the 0.17.0 docs.  The [github.io](http://googlecloudplatform.github.io/gcloud-python/stable/bigtable-usage.html) docs get updated by the same Travis job that makes the PyPI release.

**Comment by mbrukman** [comment]
Also, [RTD Bigtable usage docs](http://gcloud-python.readthedocs.io/en/stable/bigtable-usage.html) talk only about clusters, does not mention instances at all.

**Comment by tseaver** [comment]
PyPI:  https://pypi.python.org/pypi/gcloud/0.17.0
github.io: http://googlecloudplatform.github.io/gcloud-python/stable/bigtable-instance-api.html

**Comment by mbrukman** [comment]
Thank you, @tseaver!! Much appreciated.

**Pull request merged**